kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java

```diff
@@ -154,6 +162,131 @@ void shouldEncryptRecordValueForMultipleRecords() throws ExecutionException, Int
+                initial,
+                recordReceivedRecord(encrypted)).toCompletableFuture().get(10, TimeUnit.SECONDS);
+
+        record.value().rewind();
+        record2.value().rewind();
+        List<byte[]> deks = encrypted.stream()
+                .map(testingRecord -> Arrays.stream(testingRecord.headers()).filter(header -> header.key().equals("kroxylicious.io/dek")).findFirst().orElseThrow()
+                        .value())
+                .toList();
+        assertEquals(4, encrypted.size());
+        assertEquals(4, deks.size());
+        assertArrayEquals(deks.get(0), deks.get(1));
+        assertArrayEquals(deks.get(2), deks.get(3));
+        assertFalse(Arrays.equals(deks.get(0), deks.get(2)));
+        assertFalse(Arrays.equals(deks.get(0), deks.get(3)));
+    }
+
+    @Test
+    void shouldGenerateNewDekIfOldOneHasSomeRemainingEncryptionsButNotEnoughForWholeBatch() throws ExecutionException, InterruptedException, TimeoutException {
+        var kmsService = UnitTestingKmsService.newInstance();
+        InMemoryKms kms = kmsService.buildKms(new UnitTestingKmsService.Config());
+        var km = new InBandKeyManager<>(kms, BufferPool.allocating(), 3);
+
+        var kekId = kms.generateKey();
+
+        var value = ByteBuffer.wrap(new byte[]{ 1, 2, 3 });
+        TestingRecord record = new TestingRecord(value);
+
+        var value2 = ByteBuffer.wrap(new byte[]{ 3, 4, 5 });
+        TestingRecord record2 = new TestingRecord(value2);
+
+        List<TestingRecord> encrypted = new ArrayList<>();
+        List<TestingRecord> initial = List.of(record, record2);
+        km.encrypt(new EncryptionScheme<>(kekId, EnumSet.of(RecordField.RECORD_VALUE)),
+                initial,
+                recordReceivedRecord(encrypted)).toCompletableFuture().get(10, TimeUnit.SECONDS);
+        record.value().rewind();
+        record2.value().rewind();
+
+        // at this point we have encrypted 2 records with the manager set to maximum 3 encryptions per dek, so we need a new dek to encrypt 2 more records
+
+        km.encrypt(new EncryptionScheme<>(kekId, EnumSet.of(RecordField.RECORD_VALUE)),
+                initial,
+                recordReceivedRecord(encrypted)).toCompletableFuture().get(10, TimeUnit.SECONDS);
+
+        record.value().rewind();
+        record2.value().rewind();
+        List<byte[]> deks = encrypted.stream()
+                .map(testingRecord -> Arrays.stream(testingRecord.headers()).filter(header -> header.key().equals("kroxylicious.io/dek")).findFirst().orElseThrow()
+                        .value())
+                .toList();
+        assertEquals(4, encrypted.size());
+        assertEquals(4, deks.size());
+        assertArrayEquals(deks.get(0), deks.get(1));
+        assertArrayEquals(deks.get(2), deks.get(3));
+        assertFalse(Arrays.equals(deks.get(0), deks.get(2)));
+        assertFalse(Arrays.equals(deks.get(0), deks.get(3)));
```

Comment:
Would some assertJ assertions make it easier to see what properties are being validated here?

Reply:
I rejigged things a bit so that you can ask the InMemoryKms for the EDEKs it has generated in the past. From there the test can check that edeks[0] was used for records 1 and 2, and edeks[1] was used for records 3 and 4. And I've switched to assertj.